### PR TITLE
Update config for FriBiDi dynamic linking

### PR DESCRIPTION
**Pull request by nulano**
For https://github.com/python-pillow/Pillow/pull/5062.

~~See passing run in my branch: https://github.com/nulano/pillow-wheels/actions/runs/684862775~~

More comments inline below.

### CODE REVIEW

File: config.sh [hunk 1]
```diff
@@ -17,6 +18,21 @@ LIBWEBP_VERSION=1.1.0
...
+function untar {
+    local in_fname=$1
+    if [ -z "$in_fname" ];then echo "in_fname not defined"; exit 1; fi
+    local extension=${in_fname##*.}
+    case $extension in
+        tar) tar -xf $in_fname ;;
+        gz|tgz) tar -zxf $in_fname ;;
+        bz2) tar -jxf $in_fname ;;
+        zip) unzip -qq $in_fname ;;
+        xz) unxz -c $in_fname | tar -xf - ;;
```
Comment (nulano) [inline config.sh:31]: This line is wrong in the multibuild version causing an error in `tar`; will report tomorrow.

File: config.sh [hunk 2]
```diff
@@ -69,13 +85,34 @@ function pre_build {
         build_freetype
     fi
 
+    if [ -z "$IS_OSX" ]; then
+        export FREETYPE_LIBS=-lfreetype
+        export FREETYPE_CFLAGS=-I/usr/local/include/freetype2/
+    fi
```
Comment (nulano) [inline config.sh:91]: HarfBuzz fails to find FreeType on linux without this, specify the path manually.

File: config.sh [hunk 3]
```diff
@@ -69,13 +85,34 @@ function pre_build {
...
+    build_simple harfbuzz $HARFBUZZ_VERSION https://github.com/harfbuzz/harfbuzz/releases/download/$HARFBUZZ_VERSION tar.xz --with-freetype=yes --with-glib=no
```
Comment (nulano) [inline config.sh:92]: `--with-glib=no` to avoid pulling in unrelated libraries on macos.

File: config.sh [hunk 4]
```diff
@@ -69,13 +85,34 @@ function pre_build {
...
 }
 
+function pip_wheel_cmd {
+    local abs_wheelhouse=$1
+    if [ -z "$IS_OSX" ]; then
+        CFLAGS="$CFLAGS --std=c99"  # for Raqm
+    fi
+    pip wheel $(pip_opts) \
+        --global-option build_ext --global-option --enable-raqm \
+        --global-option --vendor-raqm --global-option --vendor-fribidi \
```
Comment (nulano) [inline config.sh:112]: Overriding `pip_wheel_cmd` to add the needed global options, as defined in https://github.com/python-pillow/Pillow/pull/5062.

File: config.sh [hunk 5]
```diff
@@ -84,7 +121,12 @@ function run_tests_in_repo {
 
 EXP_CODECS="jpg jpg_2000 libtiff zlib"
 EXP_MODULES="freetype2 littlecms2 pil tkinter webp"
-EXP_FEATURES="transp_webp webp_anim webp_mux xcb"
+if [ -z "$IS_OSX" ] && [[ "$MB_PYTHON_VERSION" != pypy3* ]]; then
+  EXP_FEATURES="fribidi harfbuzz raqm transp_webp webp_anim webp_mux xcb"
+else
+  # can't find FriBiDi
+  EXP_FEATURES="transp_webp webp_anim webp_mux xcb"
```
Comment (nulano) [inline config.sh:128]: macOS and PyPy on Ubuntu don't find FriBiDi preinstalled. Adding `brew install fribidi` only finds it on Python 3.6. I'm not sure why, but it should be irrelevant given the `build_ext` flags defined above.

File: config.sh [hunk 6]
```diff
@@ -69,13 +85,34 @@ function pre_build {
...
     for filename in dependency_licenses/*; do
       echo -e "\n\n----\n\n$(basename $filename | cut -f 1 -d '.')\n" | cat >> Pillow/LICENSE
       cat $filename >> Pillow/LICENSE
     done
 }
 
+function pip_wheel_cmd {
+    local abs_wheelhouse=$1
+    if [ -z "$IS_OSX" ]; then
+        CFLAGS="$CFLAGS --std=c99"  # for Raqm
```
Comment (nulano) [inline config.sh:108]: I think it is better to compile with C99 than to merge in the (poorly tested) C89 patch from https://github.com/python-pillow/Pillow/pull/5062/commits/b3cfe73854e74bc25a88f53b177713bfb63812e4.

**Comment by radarhere** [comment]
https://github.com/python-pillow/Pillow/pull/2753#issue-142603277
> We can't currently distribute a binary of Pillow with support for raqm due to license incompatibility, because while libraqm is MIT licensed, it links to GPL libraries.

@wiredfool do you recall which libraries you were discussing here?

**Comment by nulano** [comment]
> [python-pillow/Pillow#2753 (comment)](https://github.com/python-pillow/Pillow/pull/2753#issue-142603277)
> 
> > We can't currently distribute a binary of Pillow with support for raqm due to license incompatibility, because while libraqm is MIT licensed, it links to GPL libraries.
> 
> @wiredfool do you recall which libraries you were discussing here?

FriBiDi is under LGPL 2.1 or later: https://github.com/fribidi/fribidi#license

HarfBuzz is under "Old MIT": https://github.com/harfbuzz/harfbuzz/blob/master/COPYING

**Comment by nulano** [comment]
I've rerun the build in my branch of the latest version of https://github.com/python-pillow/Pillow/pull/5062 and updated the link in the top comment: https://github.com/nulano/pillow-wheels/actions/runs/684862775

Trying the 3.6 and 3.8 wheels in a fresh LinuxMint 19.1 (based on Ubuntu 18.04) VM, I get a crash after one iteration with the latest 8.1.2 wheels from PyPI, while these wheels work for at least 1000 iterations. I also had to install `libraqm0` for the PyPI wheels while these worked out of the box.
```python
from PIL import ImageFont
for i in range(1000):
    print(i)
    ImageFont.truetype("DejaVuSans", layout_engine=ImageFont.LAYOUT_RAQM).getsize("hello")
```

**Comment by wiredfool** [comment]
Getting failures now of ```
2021-03-28T13:49:28.8410816Z   error: option --enable-raqm not recognized
```

**Comment by nulano** [comment]
> Getting failures now of
> 
> ```
> 2021-03-28T13:49:28.8410816Z   error: option --enable-raqm not recognized
> ```

https://github.com/python-pillow/Pillow/pull/5062 adds new options which are required to enable the new code. These fail on the 8.1.2 tag, but are fine on the master branch (see "latest" jobs).

File: config.sh [hunk 7]
```diff
@@ -5,6 +5,7 @@ ARCHIVE_SDIR=pillow-depends-master
 
 # Package versions for fresh source builds
 FREETYPE_VERSION=2.10.4
+HARFBUZZ_VERSION=2.7.4
```
Comment (radarhere) [inline config.sh:8]:
```suggestion
HARFBUZZ_VERSION=2.8.0
```

**Pull request merged**